### PR TITLE
refactor: refactor envelope and signer to support cose

### PR DESCRIPTION
**Pull request by binbin-li**
This PR is splitted into https://github.com/notaryproject/notation-core-go/pull/59 and https://github.com/notaryproject/notation-core-go/pull/64 if you'd like to review one by one. I'll sync updates from splitted PRs here.

### What?

Background can be checked out in https://github.com/notaryproject/notation/discussions/278

1. Created a new `signature` submodule to refactor the original `signer` submodule. 
2. Created `Envelope` interface.
3. Renamed the original `SignatureProvider` to `Signer`. 
4. Created `LocalSigner` to support native signing implementation provided by underlying crypto libraries, e.g. go-cose, go-jwt.
5. Created `internal.base.Envelope` struct to replace the orginal `SignatureEnvelope` struct.
6. Other refactoring on the algorithm and keySpec definitions.

### Test?
Added corresponding unit tests.

### Notes
cose and jws folders are left empty for now which will be addressed in following PRs.

Signed-off-by: Binbin Li <libinbin@microsoft.com>

### CODE REVIEW

File: signature/algorithm.go
Comment (shizhMSFT) [inline signature/algorithm.go:65]: nit:
```suggestion
				Msg: fmt.Sprintf("rsa key size %d is not supported", bitSize),
```

File: signature/algorithm.go
Comment (shizhMSFT) [inline signature/algorithm.go:77]: nit:
```suggestion
				Msg: fmt.Sprintf("ecdsa key size %d is not supported", bitSize),
```

File: signature/algorithm.go
Comment (shizhMSFT) [inline signature/algorithm.go:81]: Should we return `UnsupportedSigningKeyError`?

File: testhelper/certificatetest.go
Comment (shizhMSFT) [inline testhelper/certificatetest.go:42]: Where is ed25519 used? It is not in the support list of notary.

File: signature/envelope.go
Comment (shizhMSFT) [inline signature/envelope.go:1]: We should document the `signature` package like
```go
// Package signature does bla bla bla...
package signature
```

File: signature/envelope.go
Comment (shizhMSFT) [inline signature/envelope.go:38]: Let's do
```suggestion
var envelopeFuncs map[string]envelopeFunc
```
and make instance on-the-fly.

File: signature/envelope.go
Comment (shizhMSFT) [inline signature/envelope.go:57]: See https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices
```suggestion
	var types []string
```

File: signature/internal/base/envelope.go
Comment (shizhMSFT) [inline signature/internal/base/envelope.go:15]: nit:
```suggestion
	signature.Envelope        // internal envelope in a specific format (e.g. COSE, JWS)
```

File: signature/internal/base/envelope.go
Comment (shizhMSFT) [inline signature/internal/base/envelope.go:13]: It is better to call that this base envelope does common validation.

File: signature/internal/base/envelope.go
Comment (shizhMSFT) [inline signature/internal/base/envelope.go:23]: Canonicalize instead of sanitize?

File: signature/internal/base/envelope.go
Comment (shizhMSFT) [inline signature/internal/base/envelope.go:68]: Looks like the payload is validate twice. Is this block of validation necessary?

File: signature/signer.go
Comment (shizhMSFT) [inline signature/signer.go:91]:
```suggestion
func (s *signer) Sign(content []byte) ([]byte, []*x509.Certificate, error) {
	return nil, nil, fmt.Errorf("local signer doesn't support sign")
}
```

File: signature/signer.go
Comment (shizhMSFT) [inline signature/signer.go:32]: It is worth calling out that the `signer` only holds the signing key. The underlying signing implementation is provided by the underlying crypto library for the specific signature format like `go-jwt` and `go-cose`.

File: signature/signer.go
Comment (shizhMSFT) [inline signature/signer.go:33]: Maybe it makes more sense to rename `signer` to `localSigner`.

File: signature/internal/base/envelope.go
Comment (JeyJeyGao) [inline signature/internal/base/envelope.go:174]:
```suggestion
// validateSigningTime checks that signing time is within the valid range of
```

File: testhelper/certificatetest.go
Comment (binbin-li) [reply testhelper/certificatetest.go:40]: Removed

File: signature/internal/base/envelope.go
Comment (binbin-li) [reply signature/internal/base/envelope.go]: Thanks for calling it out. Removed it

File: signature/internal/base/envelope.go
Comment (binbin-li) [reply signature/internal/base/envelope.go]: nice catch!

File: signature/envelope.go
Comment (binbin-li) [reply signature/envelope.go]: good to know it!

File: signature/envelope.go
Comment (chloeyin) [inline signature/envelope.go:88]:
```suggestion
// ParseEnvelope generates an envelope by given envelope bytes with specified
```

Comment (binbin-li) [reply signature/envelope.go]: thanks for catching!

File: signature/envelope.go
Comment (patrickzheng200) [inline signature/envelope.go:6]: nit: initialization

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Comment by rgnote** [comment]
@binbin-li Can you open https://github.com/binbin-li/notation-core-go/pull/1 on notation-core-go repo?

**Comment by binbin-li** [comment]
> @binbin-li Can you open [binbin-li#1](https://github.com/binbin-li/notation-core-go/pull/1) on notation-core-go repo?

Sure, I'll create one. But that PR depends on the https://github.com/notaryproject/notation-core-go/pull/59, so it will fail some check which should pass once we merge the first one.

**Comment by yizha1** [comment]
Per the https://github.com/notaryproject/notation/discussions/263 https://github.com/orgs/notaryproject/teams/notation-sub-group-maintainers - this PR is already past a week old and has two approvals from official code-reviewers, so it is eligible for merge now.

**Comment by shizhMSFT** [comment]
This PR closes #59 and #64.

**Comment by gokarnm** [comment]
@shizhMSFT @yizha1 could we revert this merge. COSE envelope support is a significant feature and it would be ideal to get reviews from multiple maintainers (as opposed to smaller changes which can be merged after a week). Both Pritesh and me have been away last week on vacation till Sept 5th. We'll be able to review this PR this week.

**Pull request merged**
